### PR TITLE
Wrap drake plan in a function, responsive imports

### LINKED ISSUE #22: Function to erase cache if new synGet() metadata or config file changes
There are a few cases where a file would change upstream of the drake workflow that would invalidate the results: 
1. A file changes on Synapse and the version was not explicitly specified in `config.yml`
2. The config file parameters are changed. 

Functions impacted: `iterative_normalization()`, `import_metadata()`, `import_counts()`, counts target, `clean_md()`

Ideas for how to remedy this:
There are only a few functions that directly depend on either a `synGet()` call or `config::get()`. 
1. Perform a check first to see if the state has changed. If there is a change, `drake::clean()` the target.
 - I'm not sure `drake::clean()` is the best function. Additionally, this approach would require explicit conditional logic. e.g. if metadata input changes; invalidate `import_metadata()` target.

### PR DESCRIPTION
**Pull request by kelshmo**
The `check_targets.R` file is meant to load `config` values in the local environment so the `drake` workflow tool can detect when an upstream target changes. `drake` caching is dictated by `memoise`. If a value in the `config.yml` file is changed and the file is saved, without the changes proposed in this PR, the workflow would not detect a change and would not re-execute the code.
Screen Shot 2020-05-22 at 4 47 31 PM

I'm not totally satisfied with this fix as I feel like it burdens the end user to remember to execute [steps](https://github.com/kelshmo/sageseqr/compare/test-config-dependency?expand=1#diff-afb8bb86d7ba66193d03700b3f562f5bR13-R16) in the `initialize.R` file multiple times and is an additional file to create once the package is installed. 

I like keeping the `config.yml` since it keeps all of the mutable values siloed. 

To test invalidating/updating `config` values, run `initialize.R` to load the workflow with test data (the current default configuration pulls down publicly available data for Synapse users.) Then change `metadata` in the `config` from `syn22090070` to `syn21072902` (it is a copied file but an updated synID should invalidate the target). Run `drake::outdated(plan)` and `drake::vis_drake_graph(plan, targets_only = TRUE)` to display the targets have correctly been tagged as "outdated".

The new `biomart synID` is to make the workflow run faster by skipping that step. 

Fixes #22 and I opened #52 to track the more sophisticated use case of identifying when a file has changed from Synapse metadata.

I would appreciate a sanity check or confirmation that I can even implement this structure with an R package.

### CODE REVIEW

**Comment by karawoo** [comment]
I don't think drake uses memoise, at least [I can't find any places](https://github.com/search?q=memoise+repo%3Aropensci%2Fdrake+extension%3A.R&type=Code&ref=advsearch&l=&l=) where it's used in the drake codebase. That said, I'm not able to test out the changes in this PR by following the steps you described -- the initialize.R script fails

```
$ Rscript initialize.R
Welcome, Kara Woo!NULL
▶ target import_metadata
Downloading  [####################]100.00%   432.0bytes/432.0bytes (101.7kB/s) clinical_metadata_copy.csv Done...    ▶ target import_counts
Downloading  [####################]100.00%   1.0MB/1.0MB (1.9MB/s) counts.csv Done...    ▶ target clean_md
▶ target counts
▶ target biomart_results
Downloading  [####################]100.00%   1.2MB/1.2MB (3.5MB/s) Ensembl-Release-xx.csv Done...    ▶ target filtered_counts
✖ fail filtered_counts
Error: target filtered_counts failed.
diagnose(filtered_counts)error$message:
  could not find function "%>%"
diagnose(filtered_counts)error$calls:
  1. └─global::filter_genes(md = clean_md, count_df = counts)
Execution halted
```

Can you say more about how you envision a user of this package interacting with this workflow as a whole? From what I can tell, after running `remotes::install_github("kelshmo/sageseqr")` and loading the package they would have to: 
1. Write an initialize.R file (since the one in this repo is in .Rbuildignore and therefore not bundled with the package itself)
2. Write a config file
3. Write a check_targets.R file
4. Write a drake plan (since the plan itself is not exported).

Is that right? If so, I do agree it's a lot of steps to keep track of.

**Comment by kelshmo** [comment]
Actually, you are correct. `drake` doesn't use `memoise`, rather models its caching mechanism around `memoise`: (doesn't link well, from the drake manual):

> Memoization is the strategic caching of the return values of functions. It is a lightweight approach to the core problem that drake and other pipeline tools are trying to solve. Every time a memoized function is called with a new set of arguments, the return value is saved for future use. Later, whenever the same function is called with the same arguments, the previous return value is salvaged, and the function call is skipped to save time. The memoise package is the primary implementation of memoization in R.

> Memoization saves time for small projects, but it arguably does not go far enough for large reproducible pipelines. In reality, the return value of a function depends not only on the function body and the arguments, but also on any nested functions and global variables, the dependencies of those dependencies, and so on upstream. drake tracks this deeper context, while memoise does not.

So that error is related to #48, I think. There are cases where the code is failing due to "missing" package dependencies, despite those dependencies being listed in the DESCRIPTION file.  `could not find function "%>%"` should be taken care of by the`magrittr` package. Am I missing something about how dependencies are dealt with in packages? 

To echo what you already stated, the steps would be: After running remotes::install_github("kelshmo/sageseqr") and loading the package usere would have to:

   1. Write an initialize.R file (since the one in this repo is in .Rbuildignore and therefore not bundled with the package itself)
   2. Write a config file
   3. Write a check_targets.R file

I'm hoping to export the plan and am working on that in a separate branch.

**Comment by karawoo** [comment]
I added a comment on #48 about the broader issue. I think the issue with `%>%` is that the initialize.R file does not load the sageseqr package, it only sources the functions, so the way you've imported the magrittr pipe [here](https://github.com/kelshmo/sageseqr/blob/64b2b433d81909f654b007a5a4cd04d3661be5d0/R/functions.R#L34) doesn't attach `%>%` to R's search path. So it seems there might be other steps I need to follow to reproduce the behavior of this PR. Should I be running `load_all()` before I run `initialize.R`.

**Comment by kelshmo** [comment]
Is it syntactically correct to call `library(sageseqr)` in `initialize.R`? I have been using `devtools::load_all()` in-practice to simulate loading the package.

I am thinking of putting `initialize.R` in the `inst` directory.

I dramatically changed this workflow - wrapping the plan and using tidy evaluation solved all my problems!!

1. Write a config file. 
2. Run`initialize.R`.

Now when the configs change, by running `rnaseq_plan` and updating the `plan` object, the tool correctly identifies outdated files. You can check with `outdated(plan)`. 

I still have errors noted with package attachments noted in #48, but I think I will resolve it outside of this PR. I need to do more research.

File: R/plan.R [hunk 1]
```diff
@@ -1,19 +1,48 @@
...
+#' @param gene_id_input Column name of the gene ids in the counts_id file.
+#' @param factor_input Vector of factor variables. Variables must be present
+#' in the metadata as column names.
+#' @param continuous_input Vector of continuous variables. Variables must be
+#' present in the metadata as column names.
+#' @param biomart_id Synapse ID to biomart object.
+#' @param biomart_version Optionally, include Synapse file version number.
+#' @inheritParams get_biomart
+#' @export
+rnaseq_plan <- function(metadata_id, metadata_version, counts_id,
```
Comment (karawoo) [inline R/plan.R:21]: I think you need to run `devtools::document()` to make sure the .Rd documentation gets generated.

File: R/plan.R [hunk 2]
```diff
@@ -1,19 +1,48 @@
...
+                        counts_version, gene_id_input, factor_input,
+                        continuous_input, gene_id,
+                        biomart_id, biomart_version, host, filters,
+                        organism){
+  drake::drake_plan(
+    import_metadata = get_data(!!metadata_id,
+                               !!metadata_version),
+    import_counts = get_data(!!counts_id,
+                             !!counts_version),
+    counts = tibble::column_to_rownames(import_counts,
```
Comment (karawoo) [inline R/plan.R:31]: R CMD check is giving a note about `import_counts` (and some of the other objects created in your plan) -- `no visible binding for global variable 'import_counts'`. I guess it doesn't know that drake will have created it by the time it's used here. It's probably not going to cause any problems so you could probably just ignore the note, or use the [workaround suggested here](https://community.rstudio.com/t/how-to-solve-no-visible-binding-for-global-variable-note/28887/2).

**Pull request merged**